### PR TITLE
Added convenience method for saving DataArray to netCDF file

### LINKED ISSUE #915: Consider adding a convenience method for saving DataArrays to netCDF files
e.g.,
```python
class DataArray:
    ...
    def to_netcdf(self, *args, **kwargs):
        dataset = self.to_dataset(name='__data__')
        data.to_netcdf(*args, **kwargs)

def open_dataarray(*args, **kwargs):
    dataset = open_dataset(*args, **kwargs)
    return dataset['__data__']
```

### PR DESCRIPTION
**Pull request by robintw**
Added a simple function to DataArray that creates a dataset with one variable
called 'data' and then saves it to a netCDF file. All parameters are passed through
to to_netcdf().

Added an equivalent function called `open_dataarray` to be used to load from these files.

Fixes #915.

### CODE REVIEW

**Comment by shoyer** [comment]
This is great, but it needs tests and documentation :).

**Comment by shoyer** [comment]
I'm also a little concerned about the design here, because it always saves/restores the DataArray with the name `'data'`, which could override the original name. Ideally this method should save/restore a DataArray as faithfully as possible.

Instead, maybe we should only override the original name if it's `None`, or the same name as one of it's coordinates (which would otherwise result in an invalid NetCDF file).

Then we could restore by looking for a Dataset with single data variable regardless of the name, e.g.,
```
if len(dataset.data_vars) != 1:
    raise ValueError
data_array, = dataset.data_vars.values()
```

**Comment by robintw** [comment]
Thanks @shoyer - I wasn't actually aware that DataArrays can have a name _before_ they're put into a Dataset (that's one of the things I love about contributing to OSS: you always learn things about the software that you never knew before!). I'll deal with the design issues ASAP.

How would you suggest testing this? I wasn't sure where to put the tests, and what exactly to test? Presumably something about the equivalence between this and saving after manually converting to a dataset, but I wasn't really sure how best to go about it.

**Comment by shoyer** [comment]
I would add these tests to test_backends, probably a new test class. The
best test is probably a round-trip test where we save and load a DataArray
and verify that the result is identical to what we started with.

**Comment by robintw** [comment]
I think I've dealt with the comments/issues now, but let me know if more needs doing.

File: xarray/backends/api.py [hunk 1]
```diff
@@ -233,6 +233,30 @@ def maybe_decode_store(store, lock=False):
...
+    """
+    Opens an `xarray.DataArray` from a netCDF file, and returns the single
+    variable that is within the file. If multiple variables are present then
+    a ValueError is raised.
+
+    This is designed to read files saved with `xarray.DataArray.to_netcdf`.
+
+    Parameters
+    ----------
+    All parameters are passed directly to `xarray.open_dataset`.
```
Comment (shoyer) [inline xarray/backends/api.py]: We should mention the parameters again (at least `path`). The fact that this uses `open_dataset` internally is an implementation detail.

File: xarray/backends/api.py [hunk 2]
```diff
@@ -233,6 +233,30 @@ def maybe_decode_store(store, lock=False):
     return maybe_decode_store(store)
 
 
+def open_dataarray(*args, **kwargs):
+    """
+    Opens an `xarray.DataArray` from a netCDF file, and returns the single
```
Comment (shoyer) [inline xarray/backends/api.py]: Keep the first line of docstrings to a single line.

File: xarray/test/test_backends.py [hunk 3]
```diff
@@ -1063,3 +1064,29 @@ def test_extract_h5nc_encoding(self):
...
+    def test_dataarray_to_netcdf_with_name(self):
+        original_da = DataArray(np.arange(12).reshape((3, 4)),
+                                name='test')
+
+        with create_tmp_file() as tmp:
+            original_da.to_netcdf(tmp)
+
+            loaded_da = open_dataarray(tmp)
+
+        self.assertDataArrayIdentical(original_da, loaded_da)
```
Comment (shoyer) [inline xarray/test/test_backends.py]: I think identical also checks name already? Worth double checking, though.

File: xarray/core/dataarray.py [hunk 4]
```diff
@@ -1095,6 +1095,31 @@ def to_masked_array(self, copy=True):
...
+        Only xarray.Dataset objects can be written to netCDF files, so
+        the xarray.DataArray is converted to a xarray.Dataset object
+        containing a single variable. If the DataArray has no name, then
+        it is given the name 'data'.
+        """
+        if not self.name and self.name not in list(self.coords):
+            # If the name isn't blank/None or the same as one of the coords
+            # (as the latter is invalid for a netCDF file) then give it a name
+            # of `data`
+            dataset = self.to_dataset(name='data')
```
Comment (shoyer) [inline xarray/core/dataarray.py]: Note that in principle "data" could also be the name of a coordinate!

I would be happier picking something long and arbitrary that a user is unlikely to use accidentally, e.g., `"__xarray_dataarray_variable__"`. Then, we could check for this name in `open_dataarray` and remove it from the returned DataArray if necessary.

File: xarray/backends/api.py [hunk 5]
```diff
@@ -233,6 +233,30 @@ def maybe_decode_store(store, lock=False):
...
+    variable that is within the file. If multiple variables are present then
```
Comment (shoyer) [inline xarray/backends/api.py]: single variable -> single data variable

File: xarray/backends/api.py [hunk 6]
```diff
@@ -233,6 +233,30 @@ def maybe_decode_store(store, lock=False):
...
+    a ValueError is raised.
+
+    This is designed to read files saved with `xarray.DataArray.to_netcdf`.
+
+    Parameters
+    ----------
+    All parameters are passed directly to `xarray.open_dataset`.
+    """
+    dataset = open_dataset(*args, **kwargs)
```
Comment (shoyer) [inline xarray/backends/api.py]: One downside of this approach is that it leaves a stranded file open without any possible way to close it.

We might want to add a `.close()` method to DataArray so we can clean this up. Or maybe just save that for a later PR.

File: xarray/core/dataarray.py (hunk 4 above)
Comment (shoyer) [inline xarray/core/dataarray.py]: To be extra thorough, we could save the original `name` as an attribute (e.g., `"__xarray_dataarray_name__"`) if it matches a coordinate, and then use that to set the name on the result. That could ensure that we always faithfully roundtrip the DataArray exactly.

**Comment by robintw** [comment]
Thanks for the useful comments.

I've added some bits to the Sphinx docs in the places that I think they should go - but let me know if I need to add any more.

**Comment by robintw** [comment]
Also, it seems like there might be an issue with not closing the Dataset - but only on Windows (see the AppVeyor build failure at https://ci.appveyor.com/project/jhamman/xarray-injyf/build/1.0.268/job/kr8ss684ijxg55be).

Any ideas how this could be sorted without having to implement `.close()` for `DataArray` (or, alternatively, how to go about implementing that, given that I don't think a DataArray contains a reference to the Dataset that it may have come from)

**Comment by shoyer** [comment]
The easiest fix would be to add a decorator to skip tests on Windows, like the existing decorators that we use:
https://github.com/pydata/xarray/blob/master/xarray/test/__init__.py#L73

I'm not sure off hand how to detect if we're running on Windows but it can't be hard.

To add a close method, I would copy these lines from `dataset.py` (or better use, move them to the common base class `BaseDataObject` in `common.py`):
https://github.com/pydata/xarray/blob/master/xarray/core/dataset.py#L251-L262

Then, you just need to define `_file_obj = None` in `DataArray.__init__` and set it when appropriate.

**Comment by robintw** [comment]
Thanks - I've got that working with the `close()` method, and I did it 'properly' using the `BaseDataObject` class.

The only failure on here now is the same as #991 - a conda error creating one of the environments.

File: xarray/core/common.py [hunk 7]
```diff
@@ -578,6 +578,19 @@ def where(self, cond, other=None, drop=False):
 
         return outobj._where(outcond)
 
+    def close(self):
+        """Close any files linked to this dataset
```
Comment (shoyer) [inline xarray/core/common.py]: dataset -> object

File: xarray/backends/api.py [hunk 8]
```diff
@@ -233,6 +233,101 @@ def maybe_decode_store(store, lock=False):
...
+                         'Please read with xarray.open_dataset and then select '
+                         'the variable you want.')
+    else:
+        data_array, = dataset.data_vars.values()
+
+    data_array._file_obj = dataset._file_obj
+
+    # Reset names if they were changed during saving
+    # to ensure that we can 'roundtrip' perfectly
+    if '__xarray_dataarray_name__' in dataset.attrs:
```
Comment (shoyer) [inline xarray/backends/api.py]: Set I would set these names as constants in this module and then import them, e.g.,
```
DATAARRAY_NAME = '__xarray_dataarray_name__'
DATAARRAY_VARIABLE = '__xarray_dataarray_variable__'
```

We have not been so strict about this, but it's usually considered bad form to write the same string literals multiple times in different places.

File: xarray/test/test_backends.py [hunk 9]
```diff
@@ -1063,3 +1064,42 @@ def test_extract_h5nc_encoding(self):
...
+@requires_netCDF4
+class TestDataArrayToNetCDF(TestCase):
+
+    def test_dataarray_to_netcdf_no_name(self):
+        original_da = DataArray(np.arange(12).reshape((3, 4)))
+
+        with create_tmp_file() as tmp:
+            original_da.to_netcdf(tmp)
+
+            loaded_da = open_dataarray(tmp)
```
Comment (shoyer) [inline xarray/test/test_backends.py]: Use the context manager to ensure things always get cleaned up (also less code), e.g.,
```
with open_dataaarray(tmp) as loaded_da:
    self.assertDataArrayIdentical(...)
```

File: xarray/test/test_backends.py [hunk 10]
```diff
@@ -1063,3 +1064,42 @@ def test_extract_h5nc_encoding(self):
...
+            self.assertDataArrayIdentical(original_da, loaded_da)
+            loaded_da.close()
+
+
+    def test_dataarray_to_netcdf_coord_name_clash(self):
+        original_da = DataArray(np.arange(12).reshape((3, 4)),
+                                dims=['x', 'y'],
+                                name='r')
```
Comment (shoyer) [inline xarray/test/test_backends.py]: I think you meant to pick a name that matches a coord, e.g., `name='x'`?

File: xarray/backends/api.py [hunk 11]
```diff
@@ -233,6 +233,101 @@ def maybe_decode_store(store, lock=False):
     return maybe_decode_store(store)
 
 
+def open_dataarray(*args, **kwargs):
```
Comment (shoyer) [inline xarray/backends/api.py]: as long as we're copying the docstring, let's copy the arguments for the function signature, too -- it helps auto-complete.

**Comment by shoyer** [comment]
I have a few final nits, but looking very nice!

**Comment by robintw** [comment]
Great, thanks. I think I've done these changes (frustratingly I can't see a way to use `*args`/`*kwargs` after using the full argument list in the function definition, so I've had to write it out. Let me know if there is a better way, but I suspect there isn't).

File: xarray/backends/api.py [hunk 12]
```diff
@@ -233,6 +233,107 @@ def maybe_decode_store(store, lock=False):
...
+    --------
+    open_dataset
+    """
+    dataset = open_dataset(filename_or_obj, group, decode_cf,
+                           mask_and_scale, decode_times,
+                           concat_characters, decode_coords, engine,
+                           chunks, lock, drop_variables)
+
+    if len(dataset.data_vars) != 1:
+        raise ValueError('Given file dataset contains more than one variable. '
```
Comment (shoyer) [inline xarray/backends/api.py]: variable -> data variable

**Comment by shoyer** [comment]
I have a couple inline comments that I think you missed?

File: xarray/backends/api.py (hunk 8 above)
Comment (shoyer) [inline xarray/backends/api.py]: see here

File: xarray/core/common.py (hunk 7 above)
Comment (shoyer) [inline xarray/core/common.py]: and here

**Comment by robintw** [comment]
Oops yes, sorry!

I'm having a bit of an issue with changing the strings to constants (which, by the way, I completely agree is how it should be done). I've defined the constants in `backends/api.py` and am then trying to import them in `core/dataarray.py`.

I am doing the import as:

`from ..backends import DATAARRAY_NAME, DATAARRAY_VARIABLE`

But this gives me the following error:

```
../../anaconda3/lib/python3.5/site-packages/py/_path/local.py:650: in pyimport
    __import__(modname)
xarray/__init__.py:3: in <module>
    from .core.extensions import (register_dataarray_accessor,
xarray/core/extensions.py:3: in <module>
    from .dataarray import DataArray
xarray/core/dataarray.py:9: in <module>
    from ..backends import DATAARRAY_NAME, DATAARRAY_VARIABLE
xarray/backends/__init__.py:8: in <module>
    from .netCDF4_ import NetCDF4DataStore
xarray/backends/netCDF4_.py:6: in <module>
    from .. import Variable
E   ImportError: cannot import name 'Variable'
```

It seems like my relative import is somehow mucking up another relative import inside the netCDF4 stuff. Do you have any ideas how I can solve this? Am I just doing my imports wrong?

**Comment by shoyer** [comment]
We have a bit of a circular imports mess :). For now, just import them at the top of the `to_netcdf` method.

**Comment by robintw** [comment]
Ah ok - that works (not entirely sure why, but it works!)

**Comment by robintw** [comment]
I realised that the reason that Travis and AppVeyor weren't running properly was because there were conflicts between this PR and master. I _think_ I've managed to rebase and push properly - but let me know if I've done anything wrong.

**Comment by shoyer** [comment]
Thanks!

**Pull request merged**